service/state/core_access.go

```diff
@@ -0,0 +1,91 @@
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/app"
+	apptypes "github.com/celestiaorg/celestia-app/x/payment/types"
+	sdk_tx "github.com/cosmos/cosmos-sdk/types/tx"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/tendermint/spm/cosmoscmd"
+	"google.golang.org/grpc"
+)
+
+// CoreAccessor implements Accessor over an RPC connection
+// with a celestia-core node.
+type CoreAccessor struct {
+	signer *apptypes.KeyringSigner
+	encCfg cosmoscmd.EncodingConfig
+
+	coreEndpoint string
+	coreConn     *grpc.ClientConn
+}
+
+// NewCoreAccessor dials the given celestia-core endpoint and
+// constructs and returns a new CoreAccessor with the active
+// connection.
+func NewCoreAccessor(
+	signer *apptypes.KeyringSigner,
+	encCfg cosmoscmd.EncodingConfig,
+	endpoint string,
+) (*CoreAccessor, error) {
+	return &CoreAccessor{
+		signer:       signer,
+		encCfg:       encCfg,
+		coreEndpoint: endpoint,
+	}, nil
+}
+
+func (ca *CoreAccessor) Start(ctx context.Context) error {
+	if ca.coreConn != nil {
+		return fmt.Errorf("core-access: already connected to core endpoint")
+	}
+	// dial given celestia-core endpoint
+	client, err := grpc.DialContext(ctx, ca.coreEndpoint, grpc.WithBlock())
+	if err != nil {
+		return err
+	}
+	ca.coreConn = client
+	return nil
+}
+
+func (ca *CoreAccessor) Stop(_ context.Context) error {
+	if ca.coreConn == nil {
+		return fmt.Errorf("core-access: no connection found to close")
+	}
+	err := ca.coreConn.Close()
+	if err != nil {
+		return err
+	}
+	ca.coreConn = nil
+	return nil
+}
+
+func (ca *CoreAccessor) Balance(ctx context.Context) (*Balance, error) {
```

Comment:
Can you add docstrings to public methods? For example, it's completely non-obvious that `Balance()` returns the balance of a particular address.

Reply:
they're documented in the interface

Reply:
Oh I see

Reply:
I am in favour of documenting them only in interface as there are less places for inconsistent docs if we do end up changing them in one place and not the other.

Reply:
I agree with you, but we should also be on the same page with @liamsi

Reply:
Hmm, can you have a look on how they look on godoc? I'd make the decision depending on that: how are the implemented methods displayed and is there a connection from the impl to the interface on godoc.

Reply:
Good point, cc @renaynay

Reply:
ref: https://github.com/celestiaorg/celestia-node/issues/426#issue-1131238570

Reply:
What was the conclusion here?

Reply:
Conclusion: if it implements an interface, those methods do not need to be documented in the implementation (only in the interface) as it reduces surface-area for having discrepancies in documentation. Any methods outside of interface impl on that struct must be documented, though.